src/backend/drm/gbm.rs

```diff
@@ -0,0 +1,336 @@
+        }
+    }
+
+    #[cfg(all(feature = "backend_egl", feature = "use_system_lib"))]
+    if matches!(
+        crate::backend::renderer::buffer_type(buffer),
+        Some(crate::backend::renderer::BufferType::Egl)
+    ) {
+        let bo = gbm
+            .import_buffer_object_from_wayland::<()>(buffer, gbm::BufferObjectFlags::SCANOUT)
+            .map_err(Error::Import)?;
+        let fb = framebuffer_from_bo_internal(
+            drm,
+            BufferObjectInternal {
+                bo: &bo,
+                offsets: None,
+                pitches: None,
+            },
+            allow_opaque_fallback,
+        )
+        .map_err(Error::Drm)?;
+
+        return Ok(Some(GbmFramebuffer {
+            _bo: Some(bo),
+            fb,
+            drm: drm.clone(),
+        }));
+    }
+
+    Ok(None)
+}
+
+/// Possible errors for attaching a [`framebuffer::Handle`]
+#[derive(Error, Debug)]
+pub enum Error {
+    /// Importing the [`Dmabuf`] to gbm failed
+    #[error("failed to import the dmabuf to gbm")]
+    Import(std::io::Error),
+    /// Failed to add a framebuffer for the bo
+    #[error("failed to add a framebuffer for the bo")]
+    Drm(DrmError),
+}
+
+/// Attach a framebuffer for a [`Dmabuf`]
+///
+/// This tries to import the [`Dmabuf`] using gbm and attach
+/// a [`framebuffer::Handle`] for the imported [`BufferObject`]
+pub fn framebuffer_from_dmabuf<A: AsFd + 'static>(
+    drm: &DrmDeviceFd,
+    gbm: &gbm::Device<A>,
+    dmabuf: &Dmabuf,
+    allow_opaque_fallback: bool,
+) -> Result<GbmFramebuffer, Error> {
+    let bo: BufferObject<()> = dmabuf
+        .import_to(gbm, gbm::BufferObjectFlags::SCANOUT)
+        .map_err(Error::Import)?;
+
+    // We override the offsets and pitches here cause the imported bo
+    // can return the wrong values. bo will only return the correct values
+    // for buffers we have allocated, but not for all client provided buffers.
+    let mut offsets: [u32; 4] = [0; 4];
+    let mut pitches: [u32; 4] = [0; 4];
+
+    for (index, offset) in dmabuf.offsets().enumerate() {
+        offsets[index] = offset;
+    }
+
+    for (index, stride) in dmabuf.strides().enumerate() {
+        pitches[index] = stride;
+    }
+
+    framebuffer_from_bo_internal(
+        drm,
+        BufferObjectInternal {
+            bo: &bo,
+            offsets: Some(offsets),
+            pitches: Some(pitches),
+        },
+        allow_opaque_fallback,
+    )
+    .map_err(Error::Drm)
+    .map(|fb| GbmFramebuffer {
+        _bo: Some(bo),
+        fb,
+        drm: drm.clone(),
+    })
+}
+
+/// Possible errors for attaching a [`framebuffer::Handle`] with [`framebuffer_from_bo`]
+#[derive(Debug, Error)]
+#[error("failed to add a framebuffer")]
+pub struct DrmError {
+    /// Error message associated to the access error
+    pub errmsg: &'static str,
+    /// Device on which the error was generated
+    pub dev: Option<PathBuf>,
+    /// Underlying device error
+    pub source: drm::SystemError,
```

Comment:
I assume thiserror has an attribute to forward Error::source to this field?

Reply:
Done